### ISSUE TITLE
[Bug]: java.lang.NullPointerException

### ISSUE THREAD

**Issue by glennj** [ballerina-platform/ballerina-lang#40354]
### Description

```
[2023-05-06 22:22:38,884] SEVERE {b7a.log.crash} - null
java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:159)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRBasicBlock.accept(BIRNode.java:474)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$3(BIROptimizer.java:121)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:121)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRFunction.accept(BIRNode.java:432)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$1(BIROptimizer.java:107)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:107)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRPackage.accept(BIRNode.java:89)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer.optimizePackage(BIROptimizer.java:83)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.genBIR(BIRGen.java:287)
        at io.ballerina.projects.internal.CompilerPhaseRunner.birGen(CompilerPhaseRunner.java:220)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performBirGenPhases(CompilerPhaseRunner.java:157)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:451)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:385)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:170)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:139)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:117)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1134)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:171)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:116)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:191)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.TestCommand.execute(TestCommand.java:315)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)

ERROR [anagram:(1:1,1:1)] Compilation failed due to an unhandled exception
error: compilation contains errors
```

### Steps to Reproduce

https://exercism.org/tracks/ballerina/exercises/anagram

code
```ballerina
public function anagrams(string word, string[] candidates) returns string[] {
    var toKey = function (string w) returns int[] => w.toCodePointInts().sort();
    var lcWord = word.toLowerAscii();
    var key = toKey(lcWord);

    return from string candidate in candidates
           let string lcCand = candidate.toLowerAscii()
           where lcCand != lcWord && key == toKey(lcCand)
           select candidate;
}
```

### Affected Version(s)

```
$ bal version
Ballerina 2201.5.0 (Swan Lake Update 5)
Language specification 2022R4
Update Tool 1.3.14
```

### OS, DB, other environment details and versions

```
$ uname -a
Linux exercism-vm 5.10.0-20-amd64 #1 SMP Debian 5.10.158-2 (2022-12-13) x86_64 GNU/Linux
```

### Related area

-> Compilation

**Comment by glennj** [comment]
A resolution:
```ballerina
public function anagrams(string word, string[] candidates) returns string[] {
    var lcWord = word.toLowerAscii();
    var key = keyOf(lcWord);

    return from string candidate in candidates
           let string lcCand = candidate.toLowerAscii()
           where lcCand != lcWord && key == keyOf(lcCand)
           select candidate;
}

function keyOf(string w) returns int[] {
    int[]|error key = w.toCodePointInts().sort();
    if key is error {
        panic(key);
    }
    return key;
}
```

**Comment by glennj** [comment]
Sorry what? That I found a workaround for a language bug makes the first attempt an "engineering mistake"? My mistake or yours? Can you point me to some place in  the documentation that would tell me the initial version is wrong?

**Comment by gimantha** [comment]
> Sorry what? That I found a workaround for a language bug makes the first attempt an "engineering mistake"? My mistake or yours? Can you point me to some place in the documentation that would tell me the initial version is wrong?

Hey @glennj The mistake is from our side. :-). Your initial code should have worked properly, but due to a mistake (Probably we have not covered your use-case in the initial testing) done from our side, you encountered this issue. Therefore we marked it as a mistake from our side and the label is only used to categorised them as an engineering mistake done from our side. We closed the issue as we fixed the issue with this pull request: #40529 

I am sorry for the confusion.
